src/sparse/impl/KokkosSparse_cluster_gauss_seidel_impl.hpp

```diff
@@ -187,14 +187,17 @@ namespace KokkosSparse{
         {
           size_type row_begin = _xadj(row);
           size_type row_end = _xadj(row + 1);
-          nnz_scalar_t sum = _Yvector(row);
-          for (size_type adjind = row_begin; adjind < row_end; ++adjind)
+          for(nnz_lno_t vec = 0; vec < (nnz_lno_t) _Xvector.extent(1); vec++)
```

Comment:
I thought this will be the inner loop, seems like it is the outer loop and we read the matrix multiple times.

Reply:
@srajama1 I would have liked to do that, but then I would need arbitrary-sized storage somewhere for the accumulated dot products in each column of X/Y. I could just allocate a view for that.

Reply:
Is it really arbitrary are is it size number of rows ? I believe reading the matrix once is going to help in terms of performance. If it is an extra array we could allocate it during symbolic and hold it in the handle.

Reply:
@srajama1 It's the number of columns in X and Y, which is arbitrary in that there's no constant upper bound. I don't know the largest it can be in practice. The idea is to allocate a view to hold the 'sum' accumulator, and this would be same dimensions as X. I couldn't allocate it during symbolic since the number of columns in X and Y isn't known until apply().

What if I just processed columns in fixed-size batches, say 8 columns at a time? At worst that's 1/8th the amount of matrix accesses as there is currently, and I wouldn't need to mess around with global or shared allocation at all.

Reply:
I like keeping it at 8 or a multiple of that. Please keep an internal variable for now.

Reply:
@brian-kelley wrote:
> What if I just processed columns in fixed-size batches, say 8 columns at a time?

I think the sparse matrix-vector multiply kernel unrolls up to 16 (or 17 -- ask @crtrott about that ;-) ).  You can decide how to do this per kernel, though it might make sense to do some measurements and figure out where the knee of the speedup curve happens (or how much the optimizations increase build times and sizes).

Reply:
@mhoemmen @srajama1 I wasn't planning on using a ``#pragma unroll`` or manual unrolling, but do you think they would help? SPMV does some extremely aggressive unrolling on CPU with fully unrolled batches of 16/17 (depending on n mod 16), and then 15 different unrolled versions in a Duff's device to deal with the remainder.

Here's what I have now - this is the RangePolicy version of apply, processing row ``ii``. apply_batch_size = 8 is constexpr. Could easily make that 16 instead.

```
          scalar_t sum[apply_batch_size];
          nnz_lno_t num_vecs = _Xvector.extent(1);
          for(nnz_lno_t batch_start = 0; batch_start < num_vecs; batch_start += apply_batch_size)
          {
            nnz_lno_t this_batch_size = apply_batch_size;
            if(batch_start + this_batch_size >= num_vecs)
              this_batch_size = num_vecs - batch_start;
            //the current batch of columns given by: batch_start, this_batch_size
            for(nnz_lno_t i = 0; i < this_batch_size; i++)
              sum[i] = _Yvector(ii, batch_start + i);
            for (size_type adjind = row_begin; adjind < row_end; ++adjind){
              nnz_lno_t colIndex = _adj(adjind);
              nnz_scalar_t val = _adj_vals(adjind);
              for(nnz_lno_t i = 0; i < this_batch_size; i++)
                sum[i] -= val * _Xvector(colIndex, batch_start + i);
            }
            nnz_scalar_t invDiagonalVal = _permuted_inverse_diagonal(ii);
            for(nnz_lno_t i = 0; i < this_batch_size; i++)
              _Xvector(ii, batch_start + i) += omega * sum[i] * invDiagonalVal;
```

Reply:
@brian-kelley I don't think you necessarily have to do that, and it could make sense for the constants to differ.  @iyamazaki reported the SGS issue [here](https://github.com/kokkos/kokkos-kernels/issues/488); I'll call him up and see if he had a particular number of columns in mind.

Reply:
@brian-kelley Hm, he doesn't have a phone number listed in the directory yet.  Might he be in CSRI today?

Reply:
@brian-kelley : The sizes we care are around 3, so this is pretty good. Whenever you are ready, we can push this in.

Reply:
Thank you! This is great!! I'll give it a try from ifpack2 and let you know.